File: python/generate_matrix.py
#!/usr/bin/env python3
import numpy as np
from scipy import sparse
from numpy.linalg import inv
from rogues import randsvd, kms, dorr, lesp # gallery() equivalent for Python

def generate_matrix(Id, N, unif_low=-1, unif_high=1):
    b_unif = np.random.uniform(unif_low, unif_high, N)
    a_unif = np.random.uniform(unif_low, unif_high, N-1)
    c_unif = np.random.uniform(unif_low, unif_high, N-1)

    # https://docs.python.org/3.10/whatsnew/3.10.html#pep-634-structural-pattern-matching
    match ID:
        case 1:
            return tridiag(a_unif, b_unif, c_unif)
        
        case 2:
            b = 1e8*np.ones(N)
            return tridiag(a_unif, b, c_unif)
        
        case 3:
            return lesp(N)

        case 4:
            a = a_unif
            a[np.floor(N/2)] = 1e-50 * a_unif[np.floor(N/2)]
            a[np.floor(N/2)+1] = 1e-50 * a_unif[np.floor(N/2)+1]

            return tridiag(a, b_unif, c_unif)

        case 5:
            # each element of a has 50% chance to be zero
            # without dtype=bool, the indexed array consists of a[0] and a[1]
            mask_a = np.array(np.random.binomial(1, 0.5, N-1), dtype=bool)
            a = np.zeros(N-1)
            a[mask_a] = a_unif[mask_a]

            # each element of c has 50% chance to be zero
            mask_c = np.array(np.random.binomial(1, 0.5, N-1), dtype=bool)
            c = np.zeros(N-1)
            c[mask_c] = c_unif[mask_c]

            return tridiag(a, b_unif, c)

        case 6:
            b = 64*np.ones(N)
            return tridiag(a_unif, b, c_unif)

        case 7:
            return inv(kms(N, rho=0.5))

        case 8:
            return randsvd(N, kappa=1e15, mode=2, kl=1, ku=1)

        case 9:
            return randsvd(N, kappa=1e15, mode=3, kl=1, ku=1)

        case 10:
            return randsvd(N, kappa=1e15, mode=1, kl=1, ku=1)

        case 11:
            return randsvd(N, kappa=1e15, mode=4, kl=1, ku=1)

        case 12:
            a = a_unif*1e-50
            return tridiag(a, b_unif, c_unif)

        case 13:
            return dorr(N, theta=1e-4)

        case 14:
            b = 1e-8*np.ones(N)
            return tridiag(a_unif, b, c_unif)

        case 15:
            b = np.zeros(N, 1)
            return tridiag(a_unif, b, c_unif)
        
        case 16:
            return tridiag(np.ones(N-1), 1e-8*np.ones(N), np.ones(N-1))

        case 17:
            return tridiag(np.ones(N-1), 1e8*np.ones(N), np.ones(N-1))

        case 18:
            return tridiag(-np.ones(N-1), 4*np.ones(N), -np.ones(N-1))

        case 19:
            return tridiag(-np.ones(N-1), 4*np.ones(N), np.ones(N-1))

        case 20:
            return tridiag(-np.ones(N-1), 4*np.ones(N), c_unif)


def tridiag(a, b, c, sparse=0):
    # Default to dense matrix
    sparse_result = 0

    # Precondition checks
    N = len(b)
    assert(len(a) == N-1)
    assert(len(c) == N-1)

    if sparse_result == 1:
        diagonals = [a, b, c]
        # https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.diags.html
        return sparse.diags(diagonals, [-1, 0, 1], shape=(N, N))
    else:
        return np.diag(a, -1) + np.diag(b) + np.diag(c, 1)

File: matlab/generate_matrix.m
%% generate_matrix
function A = generate_matrix(ID, N)
    b_unif = unifrnd(-1, 1, N, 1);
    a_unif = unifrnd(-1, 1, N-1, 1);
    c_unif = unifrnd(-1, 1, N-1, 1);
    a_norm = normrnd(0, 1, N-1, 1);
    b_norm = normrnd(0, 1, N-1, 1);
    c_norm = normrnd(0, 1, N-1, 1);

    switch ID
        case 1
            A = tridiag(a_unif, b_unif, c_unif);
        
        case 2
            b = 1e8*ones(N, 1);
            A = tridiag(a_unif, b, c_unif);
        
        case 3
            A = gallery('lesp', N);

        case 4
            a = a_unif;
            % XXX: use ceil() for 1-indexing?floor
            a(floor(N/2)) = 1e-50*a_unif(floor(N/2));
            a(floor(N/2)+1) = 1e-50*a_unif(floor(N/2)+1);

            A = tridiag(a, b_unif, c_unif);

        case 5
            % each element of a has 50% chance to be zero
            mask_a = logical(binornd(1, 0.5, N-1, 1));
            a = zeros(N-1, 1);
            a(mask_a) = a_unif(mask_a);

            % each element of c has 50% chance to be zero
            mask_c = logical(binornd(1, 0.5, N-1, 1));
            c = zeros(N-1, 1);
            c(mask_c) = c_unif(mask_c);

            A = tridiag(a, b_unif, c);

        case 6
            b = 64*ones(N, 1);
            A = tridiag(a_unif, b, c_unif);

        case 7
            A = inv(gallery('kms',  N, 0.5));

        case 8
            A = gallery('randsvd', N, 1e15, 2, 1, 1);

        case 9
            A = gallery('randsvd', N, 1e15, 3, 1, 1);

        case 10
            A = gallery('randsvd', N, 1e15, 1, 1, 1);

        case 11
            A = gallery('randsvd', N, 1e15, 4, 1, 1);

        case 12
            a = a_unif*1e-50;
            A = tridiag(a, b_unif, c_unif);

        case 13
            A = gallery('dorr', N, 1e-4);

        case 14
            b = 1e-8*ones(N, 1);
            A = tridiag(a_unif, b, c_unif);

        case 15
            b = zeros(N, 1);
            A = tridiag(a_unif, b, c_unif);
        
        case 16
            A = tridiag(ones(N-1, 1), 1e-8*ones(N, 1), ones(N-1, 1));

        case 17
            A = tridiag(ones(N-1, 1), 1e8*ones(N, 1), ones(N-1, 1));

        case 18
            A = tridiag(-ones(N-1, 1), 4*ones(N, 1), -ones(N-1, 1));

        case 19
            A = tridiag(-ones(N-1, 1), 4*ones(N, 1), ones(N-1, 1));

        case 20
            A = tridiag(-ones(N-1, 1), 4*ones(N, 1), c_unif);
        
        % additional matrices
        case 21 % 14b
            b = 1e-8*ones(N, 1);
            A = tridiag(a_norm, b, c_norm);
        
        case 22 % 14c
            b = 1e-8*ones(N, 1);
            A = tridiag(a_unif, b, c_norm);
        
        case 23 % 14d
            b = 1e-8*ones(N, 1);
            A = tridiag(a_norm, b, c_unif);

        case 24 % 14e - symmetric
            b = 1e-8*ones(N, 1);
            A = tridiag(a_unif, b, a_unif);

        case 25 % 14f - symmetric
            b = 1e-8*ones(N, 1);
            A = tridiag(a_norm, b, a_norm);

        case 26 % 14g - lower condition
            b = 1e-2*ones(N, 1);
            A = tridiag(a_unif, b, c_unif);

        case 27 % 8b
            A = gallery('randsvd', N, 1e8, 2, 1, 1);

        case 28 % 9b
            A = gallery('randsvd', N, 1e8, 3, 1, 1);

        case 29 % 10b
            A = gallery('randsvd', N, 1e8, 1, 1, 1);
        
        case 30 % 11b
            A = gallery('randsvd', N, 1e8, 4, 1, 1);

        otherwise
            error('invalid ID specified')
    end
end

%% tridiag
function A = tridiag(a, b, c, varargin)
    N = length(b);    
    if ~isvector(a) || ~isvector(b) || ~isvector(c)
        error("a, b, c must be vectors")
    end
    if length(a) ~= N-1
        error("a must have |a|-1 elements")
    end  
    if length(c) ~= N-1
        error("c must have |a|-1 elements")
    end

    A = sparse(diag(a, -1) + diag(b) + diag(c, 1));
end